examples/multiplayer-lan/rust/src/multiplayer_controller.rs

```diff
@@ -0,0 +1,303 @@
+const PORT: i32 = 8910;
+#[derive(GodotClass, Clone)]
+#[class(init)]
+pub struct PlayerData {
+    pub name: GString,
+}
+
+#[derive(GodotClass)]
+#[class(base=Control)]
+pub struct MultiplayerController {
+    #[export]
+    address: GString,
+    port: i32,
+    #[export]
+    game_scene: Gd<PackedScene>,
+    player_database: HashMap<NetworkId, PlayerData>,
+    number_of_players_loaded: u32,
+    multiplayer: OnReady<Gd<MultiplayerApi>>,
+    base: Base<Control>,
+}
+
+#[godot_api]
+impl MultiplayerController {
+    // called when a new "peer" gets connected with the server. Both client and server get notified about this
+    #[func]
+    fn on_peer_connected(&self, network_id: NetworkId) {
+        godot_print!("Player connected: {network_id}");
+    }
+
+    // called when a new "peer" gets disconnected with the server. Both client and server get notified about this
+    #[func]
+    fn on_peer_disconnected(&mut self, network_id: NetworkId) {
+        godot_print!("Player Disconnected: {network_id}");
+        self.player_database.remove(&network_id);
+        // TODO: delete player from game when player leaves lobby
+    }
+
+    // called only from client to send information to server through send_player_information
+    #[func]
+    fn on_connected_to_server(&mut self) {
+        godot_print!("Connected to Server!");
+        // send information to server
+        let username = self
+            .base()
+            .get_node_as::<LineEdit>("UsernameLineEdit")
+            .get_text();
+        let network_id = self.multiplayer.get_unique_id();
+        // server always has peer id of 1
+        self.base_mut().rpc_id(
+            1,
+            "send_player_information".into(),
+            &[Variant::from(username), Variant::from(network_id)],
+        );
+    }
+
+    // called only from clients
+    #[func]
+    fn on_connection_failed(&self) {
+        godot_print!("Couldn't Connect");
+    }
+
+    // utility function that converts our player database hashmap to a string
+    fn player_database_to_string(&self) -> String {
+        let mut string = String::from("");
+        for (network_id, data) in self.player_database.iter() {
+            let username = &data.name;
+            string.push_str(&format!(
+                "network_id: {network_id}, username: {username} \n"
+            ));
+        }
+        string
+    }
+
+    // this function should first be called by the player connecting to the server
+    // and then, the server should call this function on all the other players to propagate the information out
+    // this should result in each player having a fully populated player database containing everyone else in the lobby
+    #[rpc(any_peer)]
+    fn send_player_information(&mut self, name: GString, network_id: NetworkId) {
+        // insert new player data with network_id if it doesn't already exist
+        self.player_database
+            .entry(network_id)
+            .or_insert(PlayerData { name });
```

Comment:
Why PlayerData is a godot class at all? It is never wrapped in Gd<…> smart pointer and it is not exposed to godot anyhow.

Reply:
You're right, removed PlayerData and replaced it with a GString for player username